Brimdemons & Lobstrosities drop (slightly) useful organs

Jacquerel opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Goliaths, Legions, Watchers, and (as of recently) Bileworms all drop something vaguely useful when they die.
Brimdemons and Lobstrosities do not. This PR aims to fix that, so that there's at least some vague benefit to hunting them.

In this case it takes the form of organs you get when you butcher them, similar to the regenerative core from Legions.
As they're similar to the regenerative core, I modified the regenerative core to extend from a new common "monster core" typepath which these two new organs also extend.
Like the regenerative core, both of these items do something when used and something slightly different if you go to the effort of having someone implant them into your body. They also decay over time, and you can use stabilising serum to prevent this from happening.

https://user-images.githubusercontent.com/7483112/195967746-55a7d04d-224e-412d-aedc-3a0ec754db3d.mp4

The Rush Gland from the Lobstrosity lets you do a little impression of their charging attack, making you run very fast for a handful of seconds and ignoring slowdown effects. Unlike a lobstrosity you aren't actually built to do this so if you run into a mob you will fall over, and if you are doing this on the space station running into any dense object will also make you fall over (it shouldn't make you _too_ much of a pain for security to catch).
The idea here is that you use this to save time running back and forth from the mining base.

The Brimdust Sac from the Brimdemon covers you in exploding dust. The next three times you take Brute damage some of the dust will explode, dealing damage equal to an unupgraded PKA shot to anything near you (but not you).
If you do this in a space station not only is the damage proportionally lower (still matching the PKA), but it _does_ effect you and also it sets you on fire. You can remove the buff by showering it off.
The idea here is that you use this for minor revenge damage on enemies whose attacks you don't manage to dodge.

https://user-images.githubusercontent.com/7483112/195967811-0b362ba9-2da0-42ac-bd55-3809473cbc74.mp4

If you implant the Rush Gland then you can use it once every 3 minutes without consuming it, and the buff lasts very slightly longer. It will automatically trigger itself if your health gets low, which might be good (helps you escape a rough situation) or bad (didn't want to use it yet).

https://user-images.githubusercontent.com/7483112/195967888-f63f7cbd-60cd-4309-8004-203afc5b2153.mp4

If you implant the Brimdust Sac then you can use it once every 3 minutes to shake off cloud of dust which gives the buff to everyone nearby, if you want to kit out your miner squad. The dust cloud also makes you cough if you stand in it, and it's opaque. If you catch fire with this organ inside you and aren't in mining atmosphere then it will explode inside of your abdomen, which should probably be avoided, resultingly it is very risky to use this on the space station.

## Why It's Good For The Game

There's a small aspect of the Shaft Miner job where you're living off the land and scavenging parts off creatures you have hunted and I think that's pretty cool, so I would like some more of it.
It is also nice to mitigate the slight disappointment of killing a monster which just drops raw meat and maybe a pile of reagents for botany which you can't even pick up unless you brought a beaker with you for some reason.
I don't think people will really regularly be having these surgically implanted into them but maybe someone can think of something funny to do with a cloud of dust which makes people explode when they get punched some day.
Ashwalker content...?

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
add: You can harvest useful organs from Brimdemons and Lobstrosities.
fix: You can no longer 'preserve' regenerative cores which have already expired to repair them.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Jacquerel committed:
Refactors regenerative cores to extend from a new parent for other monster organ items.

Jacquerel committed:
Adds more generic behaviour to the base class.
Adds lobster rush gland.

Jacquerel committed:
Adds possibility to not consume organ.
Fixes implanted dash buff.

Jacquerel committed:
Redesign to allow persistent organs, starting to wonder if these didn't need to share a typepath after all, we'll see.

Jacquerel committed:
Finished rush gland aside from aesthetics
Begin implementing brimdust sac

Jacquerel committed:
Completed Brimdust Sac.
Added to butcher lists.
Things which could store regen core now store monster cores in general.
Added to random creature organ drop table.

Jacquerel committed:
Adds visuals.

Jacquerel committed:
Fixes lobster alert corners.

Jacquerel committed:
Remove silly comment.
Refactor buff stack desc setting.
Adds sound to explosion, silly thing to almost forget.

Jacquerel committed:
Removes accidentally duplicated movement speed effect

Kapu1178 commented on code/modules/mining/equipment/monster_organs/monster_organ.dm:
```diff
@@ -0,0 +1,202 @@
+	return ..()
+
+/obj/item/organ/internal/monster_core/Insert(mob/living/carbon/target_carbon, special = 0, drop_if_replaced = TRUE)
+	. = ..()
```
pretty sure you need an `if(!.) return` here, dont remember off the top of my head

Kapu1178 commented on code/modules/mining/equipment/monster_organs/monster_organ.dm:
```diff
@@ -0,0 +1,202 @@
+/obj/item/organ/internal/monster_core/attack_self(mob/user)
+	if (!user.canUseTopic(src, be_close = TRUE, no_dexterity = FALSE, no_tk = TRUE))
+		return
+	try_apply(user, user)
```
you know, at some point you gotta ask, why is this even an organ.

Jacquerel replied:
You're not really wrong that probably the instances of people implanting even the existing item as an organ happens one round in a hundred (even though as you noticed, it provides a full admin heal!) as opposed to just using it inhand as soon as you see it on the floor
I think that for the sandbox it's still fun to have the option to see a weird alien organ and have the ability to ask a doctor to put it into your body though

Kapu1178 commented on code/modules/mining/equipment/monster_organs/monster_organ.dm:
```diff
@@ -0,0 +1,202 @@
+	if (inert)
+		return FALSE
+	inert = TRUE
+	decay_timer = null
```
deltimer?

Kapu1178 commented on code/modules/mining/equipment/monster_organs/regenerative_core.dm:
```diff
@@ -0,0 +1,52 @@
+		activate_implanted()
+
+/obj/item/organ/internal/monster_core/regenerative_core/activate_implanted()
+	owner.revive(full_heal = TRUE, admin_revive = FALSE)
```
wow thats kinda busted

Kapu1178 commented on code/modules/mining/equipment/monster_organs/rush_gland.dm:
```diff
@@ -0,0 +1,86 @@
+	desc_inert = "A lobstrosity's adrenal gland. It is all shrivelled up."
+	user_status = /datum/status_effect/lobster_rush
+
+#define HEALTH_DANGER_ZONE 30
```
put defines at the file header and undef at footer, nesting it between procs is basically pointless.

Kapu1178 commented on code/modules/mining/equipment/monster_organs/rush_gland.dm:
```diff
@@ -0,0 +1,86 @@
+/datum/status_effect/lobster_rush/proc/smack_into(mob/living/target)
+	target.Knockdown(5 SECONDS)
+	target.adjustStaminaLoss(40)
+	target.adjustBruteLoss(20)
```
```suggestion
	target.apply_damage(20, BRUTE)
```
adjustBruteLoss() ignores armor. There's other args to apply_damage you will want to use here, i do not remember the exact order.

spread_damage should be TRUE tho

Kapu1178 commented on code/modules/mining/lavaland/tendril_loot.dm:
```diff
@@ -624,7 +624,17 @@
 	resistance_flags = FIRE_PROOF
 	clothing_flags = THICKMATERIAL
-	allowed = list(/obj/item/flashlight, /obj/item/tank/internals, /obj/item/pickaxe, /obj/item/spear, /obj/item/organ/internal/regenerative_core/legion, /obj/item/knife, /obj/item/kinetic_crusher, /obj/item/resonator, /obj/item/melee/cleaving_saw)
+	allowed = list(
+		/obj/item/flashlight,
+		/obj/item/tank/internals,
+		/obj/item/pickaxe,
+		/obj/item/spear,
+		/obj/item/organ/internal/monster_core,
+		/obj/item/knife,
+		/obj/item/kinetic_crusher,
+		/obj/item/resonator,
+		/obj/item/melee/cleaving_saw,
+		)
```
```suggestion
	allowed = list(
		/obj/item/flashlight,
		/obj/item/tank/internals,
		/obj/item/pickaxe,
		/obj/item/spear,
		/obj/item/organ/internal/monster_core,
		/obj/item/knife,
		/obj/item/kinetic_crusher,
		/obj/item/resonator,
		/obj/item/melee/cleaving_saw,
	)
```
formatting nitpick

Kapu1178 commented on code/modules/mining/equipment/monster_organs/monster_organ.dm:
```diff
@@ -0,0 +1,202 @@
+	slot = ORGAN_SLOT_MONSTER_CORE
+	organ_flags = NONE
+	force = 0
+	/// Set to true if this organ has decayed into uselessness.
+	var/inert = FALSE
+	/// ID of the timer which will decay this organ
+	var/decay_timer
+	/// Time after which organ should become useless
+	var/time_to_decay = 4 MINUTES
+	/// Icon state to apply when preserved
+	var/icon_state_preserved
+	/// Description to use once organ has been preserved
+	var/desc_preserved
+	/// Icon state to apply when inert
+	var/icon_state_inert
+	/// Description to use once organ is inert
+	var/desc_inert
+	/// Status effect applied by this organ
+	var/datum/status_effect/user_status
+	/// Moodlet applied by this organ
+	var/datum/mood_event/moodlet
+	/// Action to grant when organ is implanted
+	var/datum/action/item_action/organ_action/use_internal
```
If you're going to add this much boilerplate, don't add it to a random subtype. Preservation should be an `/organ/internal` mechanic, as should `moodlet on insert` and `action to grant`

rasepretep commented:
I like this not as a miner, but as someone who likes to play Ashwalker or go down to live as an Ashssistant. This'll make both of these a lot more fun.

Jacquerel commented on code/modules/mining/equipment/monster_organs/brimdust_sac.dm:
```diff
@@ -0,0 +1,174 @@
+	tick_interval = -1
+	consumed_on_threshold = FALSE
+	alert_type = /atom/movable/screen/alert/status_effect/brimdust_coating
+	/// Damage to deal on explosion
+	var/static/blast_damage = 40
+	/// Damage reduction when not in a mining pressure area
+	var/static/pressure_modifier = 0.25
```
Honestly cannot tell you why I did that

Jacquerel committed:
Quick formatting and oversight fixes.

ATH1909 commented:
rush gland boost should apply a positive modifier for tackles, imo

JohnFulpWillard commented:
Does this have any purpose on lavaland or does it only benefit PvP? Being immediately sent to stamcrit mid-megafauna doesn't sound like something miners would want.

Jacquerel commented:
> Does this have any purpose on lavaland or does it only benefit PvP? Being immediately sent to stamcrit mid-megafauna doesn't sound like something miners would want.

Moving faster is useful basically anywhere, especially if you're actually mining (I know this might be an unlikely suggestion). 
The speed boost isn't really meant to be used as an attack because it's as debilitating to you as your target and does (as you have identified) basically nothing to megafauna, but they pretty much all have a place where a sudden boost of speed would be convenient.
I can probably make it add a trait which makes tackles better for the very short period of time it is active though.

The retaliation damage is very dangerous to yourself on the space station as a way to try and make it less useful for PvP unless you're already fighting a miner in the place where they're strongest, whereas you're pretty likely to take incidental melee attacks from mining mobs every so often. Not particularly useful on megafauna because ideally you're not getting damaged by them at all but if you are, damage is damage.

TheGreatValvatorez commented:
brim gland pls

Shadowflame909 commented:
Yeah getting stunned while trying to run away from a mega fauna (owie they hit you) doesn't sound too good and might end up being counterintuitive/unused like the cursed heart. It's well balanced for the station though!

optimumtact commented:
> Yeah getting stunned while trying to run away from a mega fauna (owie they hit you) doesn't sound too good and might end up being counterintuitive/unused like the cursed heart. It's well balanced for the station though!

This encapsulates everything wrong with lavaland in a single comment lol

termaus commented:
I dont understand, isnt this basically jump boots with major debuffs with slightly longer range? I think this item would be greatly unbalanced, a suggestion is significanly longer duration for lobstrosity gland with something like %5 chance to drop

Jacquerel committed:
Makes some behaviour which wasn't actually generic not generic
Just use cooldown actions instead of organ actions (bonus: now we get tooltips)
Grant these using the existing atom action granting handler

Jacquerel committed:
Run the integration tests again.

MrMelbert reviewed (approved):
Code looks good

Sealed101 commented:
gonna leave this here

![image](https://user-images.githubusercontent.com/75863639/197597975-1983923b-625c-47ff-a39b-93a0b3806b47.png)

Jacquerel committed:
Merge branch 'master' into on-site-procurement

Jacquerel committed:
Replaces a new instance of hivelord stabiliser.

Fikou commented on code/modules/mining/equipment/monster_organs/rush_gland.dm:
```diff
@@ -0,0 +1,98 @@
+	if (owner.health <= HEALTH_DANGER_ZONE)
+		trigger_interal_action()
+
+/obj/item/organ/internal/monster_core/rush_gland/trigger_interal_action()
```
internal?

Jacquerel replied:
The proc name is meant to differentiate "thing it does if you squeeze it" vs "thing it does if you have it inside your body".
But maybe "organ action" or something might be clearer.

Fikou commented on code/modules/mining/equipment/monster_organs/monster_organ.dm:
```diff
@@ -0,0 +1,211 @@
+	desc = "All that remains of a monster. This abstract item should not spawn. \
+		It will rapidly decay into uselessness. but don't worry because it's already useless."
+	icon = 'icons/obj/medical/organs/mining_organs.dmi'
+	icon_state = "roro core 2"
```
rename this to hivelord_core, roros dont exist anymore

Fikou commented on code/modules/mining/equipment/monster_organs/monster_organ.dm:
```diff
@@ -0,0 +1,211 @@
+		return FALSE
+	inert = TRUE
+	deltimer(decay_timer)
+	decay_timer = null
```
why are you setting the timer to null if you already deltimered it

LarissaMayrink commented:
Please don't give miners a free regenerating meth implant from one of the weakest mobs on lavaland

Jacquerel commented:
> Please don't give miners a free regenerating meth implant from one of the weakest mobs on lavaland

While I am happy to make changes, the implanted action is active for 5 seconds every 3 minutes and if you use it on the space station you have to be very careful not to end it early by running into something and immobilising yourself for just as long as the buff would have lasted (for instance, closed doors). 
And this only after you can convince someone to surgically insert it into your body, which certainly isn't impossible but is a sufficient risk that it is a deterrent to most people implanting other useful equipment.

Jacquerel committed:
Renames icon states.
Renames proc.

LarissaMayrink commented:
>And this only after you can convince someone to surgically insert it into your body

lrp players view of the game. Also even on bagil which is wilder than sybil you can get implants done easily, you just gotta be nice. 

I'm a sucker for teethpills and have gotten one on literally every bagil round i've played in the last two weeks, besides a lot of doctors are always taking any chance to do surgery

Fikou reviewed (changes requested):
make icon names of hivelords and legions consistent with the other 2, and also set the hivelord's decaying icon

Jacquerel committed:
Renames icon states
Hivelord core now uses a different appearance before it decays, as the previous stable state is more fitting as the decayed one
